Source/WebGPU/WGSL/Overload.cpp

```diff
@@ -0,0 +1,534 @@
+/*
+ * Copyright (c) 2023 Apple Inc. All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY APPLE INC. ``AS IS'' AND ANY
+ * EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL APPLE INC. OR
+ * CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+ * EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+ * PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+ * PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
+ * OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "config.h"
+#include "Overload.h"
+
+#include <wtf/DataLog.h>
+
+namespace WGSL {
+
+static constexpr bool shouldDumpOverloadDebugInformation = false;
+static constexpr ASCIILiteral logPrefix = "> overload: "_s;
+
+template<typename... Arguments>
+inline void log(Arguments&&... arguments)
+{
+    if (shouldDumpOverloadDebugInformation)
+        dataLog(logPrefix, std::forward<Arguments>(arguments)...);
+}
+
+template<typename... Arguments>
+inline void logLn(Arguments&&... arguments)
+{
+    if (shouldDumpOverloadDebugInformation)
+        dataLogLn(logPrefix, std::forward<Arguments>(arguments)...);
+}
+
+struct ViableOverload {
+    const OverloadCandidate* candidate;
+    FixedVector<unsigned> cost;
+    Type* result;
+};
+
+class OverloadResolver {
+public:
+    OverloadResolver(TypeStore&, const WTF::Vector<OverloadCandidate>&, const WTF::Vector<Type*>&, unsigned numberOfTypeSubstitutions, unsigned numberOfValueSubstitutions);
+
+    Type* resolve();
+
+private:
+    static constexpr unsigned s_noConversion = std::numeric_limits<unsigned>::max();
+
+    FixedVector<std::optional<ViableOverload>> considerCandidates();
+    std::optional<ViableOverload> considerCandidate(const OverloadCandidate&);
+    unsigned calculateCost(const AbstractType&, Type*);
+    unsigned conversionCost(Type*, Type*) const;
```

Comment:
WGSL types don't have implicit conversions, except for literals, so I'm not sure this is the right approach.

Reply:
This is supposed to match 1:1 the "Conversion Rank" table in the spec: https://www.w3.org/TR/WGSL/#conversion-rank

Reply:
I should rename this to `conversionRank` if we end up landing this.